app/src/main/java/com/example/plantsapp/data/firebase/repository/FirebasePlantsRepository.kt

```diff
@@ -0,0 +1,83 @@
+package com.example.plantsapp.data.firebase.repository
+
+import com.example.plantsapp.data.firebase.entity.FirebasePlant
+import com.example.plantsapp.domain.model.Plant
+import com.example.plantsapp.domain.repository.PlantsRepository
+import com.google.firebase.firestore.DocumentReference
+import com.google.firebase.firestore.FirebaseFirestore
+import com.google.firebase.firestore.ktx.toObject
+import com.google.firebase.firestore.ktx.toObjects
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.tasks.await
+import timber.log.Timber
+import javax.inject.Inject
+
+class FirebasePlantsRepository @Inject constructor(
+    private val firestore: FirebaseFirestore
+) : PlantsRepository {
+
+    private val plantsCollection = firestore.collection(KEY_COLLECTION_PLANTS)
+
+    override fun observePlants(): Flow<List<Plant>> {
+        return callbackFlow {
+            plantsCollection
+                .addSnapshotListener { value, error ->
+                    when {
+                        error != null -> {
+                            Timber.e("Listen failed..", error)
+                        }
+
+                        value != null -> {
+                            val plants = value
+                                .toObjects<FirebasePlant>()
+                                .map { it.toPlant() }
+                            trySend(plants)
+                        }
+                    }
+                }
+
+            awaitClose { cancel() }
```

Comment:
Without this line throws this error:
![Screen Shot 2021-12-02 at 5 38 06 PM](https://user-images.githubusercontent.com/43578259/144443030-da8a8e6c-bc19-485a-8e8f-17b918d59cf8.png)

![Screen Shot 2021-12-02 at 5 44 31 PM](https://user-images.githubusercontent.com/43578259/144444692-4aa2fa1a-105f-4cbb-8398-3a8502188511.png)
from https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/callback-flow.html

Reply:
Good that you found it: 
So how about process this "closing" correctly?
I'm talking about cancelling this snapshot listener